### PR TITLE
startup-monitor: releases the lock on each iteration to prevent fight with the installer

### PR DESCRIPTION
**Pull request by p0lyn0mial**
additionally, it doesn't fall back when the target revision doesn't match the expected one.
this usually means something bad happened, in that case, we should report an error (the startup controller will put an operator into a degraded state)

### CODE REVIEW

File: pkg/operator/staticpod/startupmonitor/monitor.go [hunk 1]
```diff
@@ -106,18 +106,22 @@ func (m *monitor) Run(ctx context.Context, installerLock Locker) (ready bool, re
 	// we use wait.Until because it uses sliding interval, wait.Poll does not.
 	withTimeoutCtx, cancel := context.WithTimeout(ctx, m.timeout-m.probeInterval) // last iteration is done manually to have control over the Unlock
 	wait.UntilWithContext(withTimeoutCtx, func(ctx context.Context) {
-		defer func() {
-			if lastReady {
-				installerLock.Unlock()
-			}
-		}()
+		defer installerLock.Unlock()
```
Comment (sttts) [inline pkg/operator/staticpod/startupmonitor/monitor.go:109]: no, this breaks the contract.

Comment (sttts) [reply pkg/operator/staticpod/startupmonitor/monitor.go:109]: Only on error or timeout we are allowed to release the lock. If the target is ready here in the loop, we have to keep the lock kept.

File: pkg/operator/staticpod/startupmonitor/monitor.go [hunk 2]
```diff
@@ -106,18 +106,22 @@ func (m *monitor) Run(ctx context.Context, installerLock Locker) (ready bool, re
...
+	// do last iteration without calling Unlock()
+	// note that you cannot logically have lastReady = true and lastError != nil
 	if !lastReady {
-		iteration(ctx) // do last iteration without Unlock
+		iteration(ctx)
+		if lastError != nil {
+			// release the lock since we are exiting anyway
+			installerLock.Unlock()
+			return false, lastReason, lastMessage, lastError
+		}
```
Comment (sttts) [inline pkg/operator/staticpod/startupmonitor/monitor.go:124]: this case was missing :+1:

Can we test it?

File: pkg/operator/staticpod/startupmonitor/monitor.go [hunk 3]
```diff
@@ -143,7 +147,8 @@ func (m *monitor) isReady(ctx context.Context) (ready bool, reason string, messa
 		return false, "", "", err
 	}
 	if m.revision != currentTargetRevision {
-		return false, "UnexpectedRevision", fmt.Sprintf("expected revision %d, found %d", m.revision, currentTargetRevision), nil
+		// return an error so that we won't fallback when we fail to monitor unexpected target
+		return false, "", "", fmt.Errorf("expected revision %d, found %d", m.revision, currentTargetRevision)
```
Comment (sttts) [inline pkg/operator/staticpod/startupmonitor/monitor.go:151]: good point

File: pkg/operator/staticpod/startupmonitor/monitor.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/startupmonitor/monitor.go]: fixed.

File: pkg/operator/staticpod/startupmonitor/monitor.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/startupmonitor/monitor.go]: added unit tests

**Pull request merged**